src/proof/dot/dot_printer.cpp

```diff
@@ -25,55 +25,103 @@
-       pos += rep.length())
+  std::string newS;
+  newS.reserve(s.size());
+  for (const char c : s)
   {
-    s.replace(pos, rep.length() - 1, rep);
+    switch (c)
+    {
+      case '\"': newS += "\\\""; break;
+      case '>': newS += "\\>"; break;
+      case '<': newS += "\\<"; break;
+      case '{': newS += "\\{"; break;
+      case '}': newS += "\\}"; break;
+      case '|': newS += "\\|"; break;
+      default: newS += c; break;
+    }
   }
+
+  return newS;
 }
 
 void DotPrinter::print(std::ostream& out, const ProofNode* pn)
 {
   uint64_t ruleID = 0;
 
-  out << "digraph proof {\n";
-  DotPrinter::printInternal(out, pn, ruleID);
+  out << "digraph proof {\n\trankdir=\"BT\";\n\tnode [shape=record];\n";
+  DotPrinter::printInternal(out, pn, ruleID, 0, false);
   out << "}\n";
 }
 
 void DotPrinter::printInternal(std::ostream& out,
                                const ProofNode* pn,
-                               uint64_t& ruleID)
+                               uint64_t& ruleID,
+                               int scopeNumber,
+                               bool inPropositionalVision)
 {
   uint64_t currentRuleID = ruleID;
-  PfRule r = pn->getRule();
+  const std::vector<std::shared_ptr<ProofNode>>& children = pn->getChildren();
+  std::ostringstream currentArguments, resultStr, classes, colors;
 
-  out << "\t\"" << currentRuleID << "\" [ shape = \"box\", label = \"" << r;
+  out << "\t" << currentRuleID << " [ label = \"{";
 
-  std::ostringstream currentArguments, resultStr;
-  const std::vector<std::shared_ptr<ProofNode>>& children = pn->getChildren();
-  DotPrinter::ruleArguments(currentArguments, pn);
-  // guarantee that arguments do not have unescaped quotes
-  std::string astring = currentArguments.str();
-  cleanQuotes(astring);
-  out << astring << "\"];\n\t\"" << currentRuleID
-      << "c\" [ shape = \"ellipse\", label = \"";
-  // guarantee that conclusion does not have unescaped quotes
   resultStr << pn->getResult();
-  astring = resultStr.str();
-  cleanQuotes(astring);
+  std::string astring = resultStr.str();
+  out << sanitizeString(astring);
+
+  PfRule r = pn->getRule();
+  DotPrinter::ruleArguments(currentArguments, pn);
+  astring = currentArguments.str();
+  out << "|" << r << sanitizeString(astring) << "}\"";
+  classes << ", class = \"";
+  colors << "";
 
-  out << astring << "\" ];\n\t\"" << currentRuleID << "\" -> \""
-      << currentRuleID << "c\";\n";
+  switch (r)
+  {
+    case PfRule::SCOPE:
+      if (scopeNumber < 1)
+      {
+        classes << " basic";
+        colors << ", color = blue ";
+        inPropositionalVision = true;
+      }
+      scopeNumber++;
+      break;
+    case PfRule::ASSUME:
+      if (scopeNumber < 2)
+      {
+        classes << " basic";
+        colors << ", color = blue ";
+      }
+      if (inPropositionalVision)
+      {
+        classes << " propositional";
+        colors << ", fillcolor=aquamarine4, style=filled ";
+      }
```

Comment:
Are these two supposed to be exclusive (one or the other)? It's a bit confusing as it is.

Reply:
They are not supposed to be exclusive. A node can belong at the same time to the basic vision and the propositional vision, right?

Reply:
They can indeed. Add a comment to say that this is the case.